File: SlotMachineSuite/SlotMachineSuite/AttributionManager.m
#import "AttributionManager.h"
#import <objc/message.h>

static id AFSharedInstance(void) {
    Class appsFlyerClass = NSClassFromString(@"AppsFlyerLib");
    SEL sharedSelector = NSSelectorFromString(@"shared");
    if (![appsFlyerClass respondsToSelector:sharedSelector]) {
        return nil;
    }

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
    return [appsFlyerClass performSelector:sharedSelector];
#pragma clang diagnostic pop
}

@implementation AttributionManager {
    BOOL _didConfigure;
}

+ (instancetype)sharedManager {
    static AttributionManager *manager;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        manager = [[AttributionManager alloc] init];
    });
    return manager;
}

- (void)configure {
    if (_didConfigure) {
        return;
    }

    _didConfigure = YES;
    [self configureAppsFlyer];
    [self configureAdjust];
}

- (void)logEvent:(NSString *)name values:(NSDictionary<NSString *,id> *)values {
    if (name.length == 0) {
        return;
    }

    id appsFlyer = AFSharedInstance();
    SEL logSelector = NSSelectorFromString(@"logEvent:withValues:");
    if ([appsFlyer respondsToSelector:logSelector]) {
        ((void (*)(id, SEL, NSString *, NSDictionary *))objc_msgSend)(appsFlyer, logSelector, name, values ?: @{});
    }

    NSString *adjustToken = values[@"adjust_event_token"];
    if (adjustToken.length > 0) {
        Class eventClass = NSClassFromString(@"ADJEvent");
        Class adjustClass = NSClassFromString(@"Adjust");
        SEL initSelector = NSSelectorFromString(@"initWithEventToken:");
        SEL trackSelector = NSSelectorFromString(@"trackEvent:");
        if (eventClass && [adjustClass respondsToSelector:trackSelector]) {
            id event = [[eventClass alloc] init];
            if ([event respondsToSelector:initSelector]) {
                event = ((id (*)(id, SEL, NSString *))objc_msgSend)(event, initSelector, adjustToken);
                ((void (*)(id, SEL, id))objc_msgSend)(adjustClass, trackSelector, event);
            }
        }
    }
}

- (void)handleOpenURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
    id appsFlyer = AFSharedInstance();
    SEL selector = NSSelectorFromString(@"handleOpenUrl:options:");
    if ([appsFlyer respondsToSelector:selector]) {
        ((void (*)(id, SEL, NSURL *, NSDictionary *))objc_msgSend)(appsFlyer, selector, url, options ?: @{});
    }
}

- (BOOL)continueUserActivity:(NSUserActivity *)userActivity
          restorationHandler:(void (^)(NSArray<id<UIUserActivityRestoring>> * _Nullable))restorationHandler {
    id appsFlyer = AFSharedInstance();
    SEL selector = NSSelectorFromString(@"continueUserActivity:restorationHandler:");
    if ([appsFlyer respondsToSelector:selector]) {
        return ((BOOL (*)(id, SEL, NSUserActivity *, id))objc_msgSend)(appsFlyer, selector, userActivity, restorationHandler);
    }
    return NO;
}

- (void)configureAppsFlyer {
    NSString *devKey = [self infoString:@"AppsFlyerDevKey"];
    NSString *appID = [self infoString:@"AppsFlyerAppID"];
    id appsFlyer = AFSharedInstance();

    if (devKey.length == 0 || appID.length == 0 || !appsFlyer) {
        NSLog(@"[AppsFlyer] configure skipped: missing keys or framework.");
        return;
    }

    SEL initSelector = NSSelectorFromString(@"initializeWithDevKey:appId:");
    if ([appsFlyer respondsToSelector:initSelector]) {
        ((void (*)(id, SEL, NSString *, NSString *))objc_msgSend)(appsFlyer, initSelector, devKey, appID);
    }

#if DEBUG
    if ([appsFlyer respondsToSelector:NSSelectorFromString(@"setIsDebug:")]) {
        ((void (*)(id, SEL, BOOL))objc_msgSend)(appsFlyer, NSSelectorFromString(@"setIsDebug:"), YES);
    }
#endif

    if ([appsFlyer respondsToSelector:NSSelectorFromString(@"start")]) {
        ((void (*)(id, SEL))objc_msgSend)(appsFlyer, NSSelectorFromString(@"start"));
    }
}

- (void)configureAdjust {
    NSString *appToken = [self infoString:@"AdjustAppToken"];
    if (appToken.length == 0) {
        NSLog(@"[Adjust] configure skipped: missing AdjustAppToken.");
        return;
    }

    Class configClass = NSClassFromString(@"ADJConfig");
    Class adjustClass = NSClassFromString(@"Adjust");
    SEL initSelector = NSSelectorFromString(@"initWithAppToken:environment:");
    SEL initSdkSelector = NSSelectorFromString(@"initSdk:");

    if (!configClass || ![adjustClass respondsToSelector:initSdkSelector]) {
        NSLog(@"[Adjust] configure skipped: framework unavailable.");
        return;
    }

    NSString *environment = [[[self infoString:@"AdjustEnvironment"] lowercaseString] isEqualToString:@"sandbox"] ? @"sandbox" : @"production";
    id config = [[configClass alloc] init];
    if ([config respondsToSelector:initSelector]) {
        config = ((id (*)(id, SEL, NSString *, NSString *))objc_msgSend)(config, initSelector, appToken, environment);
    }

    if (config) {
        ((void (*)(id, SEL, id))objc_msgSend)(adjustClass, initSdkSelector, config);
    }
}

- (NSString *)infoString:(NSString *)key {
    id value = [NSBundle.mainBundle objectForInfoDictionaryKey:key];
    return [value isKindOfClass:NSString.class] ? value : @"";
}

@end

File: SlotMachineSuite/SlotMachineSuite/AttributionManager.swift
import Foundation
import SwiftUI
import UIKit

#if canImport(AppsFlyerLib)
import AppsFlyerLib
#endif

#if canImport(AdjustSdk)
import AdjustSdk
#elseif canImport(Adjust)
import Adjust
#endif

final class AppDelegate: NSObject, UIApplicationDelegate {
    func application(
        _ application: UIApplication,
        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil
    ) -> Bool {
        AttributionManager.shared.configure()
        return true
    }

    func application(
        _ application: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        AttributionManager.shared.handleOpenURL(url, options: options)
        return true
    }

    func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool {
        AttributionManager.shared.continueUserActivity(userActivity, restorationHandler: restorationHandler)
    }
}

final class AttributionManager {
    static let shared = AttributionManager()

    private var didConfigure = false

    private init() {}

    func configure() {
        guard !didConfigure else { return }
        didConfigure = true

        configureAppsFlyer()
        configureAdjust()
    }

    func logEvent(_ name: String, values: [String: Any] = [:]) {
        guard !name.isEmpty else { return }

#if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().logEvent(name, withValues: values)
#endif

#if canImport(AdjustSdk) || canImport(Adjust)
        if let eventToken = values["adjust_event_token"] as? String, !eventToken.isEmpty {
            let event = ADJEvent(eventToken: eventToken)
            Adjust.trackEvent(event)
        }
#endif
    }

    func handleOpenURL(_ url: URL, options: [UIApplication.OpenURLOptionsKey: Any]) {
#if canImport(AppsFlyerLib)
        AppsFlyerLib.shared().handleOpen(url, options: options)
#endif
    }

    func continueUserActivity(
        _ userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool {
#if canImport(AppsFlyerLib)
        return AppsFlyerLib.shared().continue(userActivity, restorationHandler: restorationHandler)
#else
        return false
#endif
    }

    private func configureAppsFlyer() {
#if canImport(AppsFlyerLib)
        let devKey = infoString("AppsFlyerDevKey")
        let appID = infoString("AppsFlyerAppID")

        guard !devKey.isEmpty, !appID.isEmpty else {
            NSLog("[AppsFlyer] configure skipped: missing AppsFlyerDevKey or AppsFlyerAppID in Info.plist.")
            return
        }

        let appsFlyer = AppsFlyerLib.shared()
        appsFlyer.initialize(devKey: devKey, appId: appID)
#if DEBUG
        appsFlyer.isDebug = true
#endif
        appsFlyer.start()
#endif
    }

    private func configureAdjust() {
#if canImport(AdjustSdk) || canImport(Adjust)
        let appToken = infoString("AdjustAppToken")
        guard !appToken.isEmpty else {
            NSLog("[Adjust] configure skipped: missing AdjustAppToken in Info.plist.")
            return
        }

        let environmentValue = infoString("AdjustEnvironment").lowercased()
        let environment = environmentValue == "sandbox" ? ADJEnvironmentSandbox : ADJEnvironmentProduction
        guard let config = ADJConfig(appToken: appToken, environment: environment) else {
            NSLog("[Adjust] configure skipped: invalid Adjust app token.")
            return
        }

#if DEBUG
        config.logLevel = ADJLogLevelVerbose
#else
        config.logLevel = ADJLogLevelSuppress
#endif
        Adjust.initSdk(config)
#endif
    }

    private func infoString(_ key: String) -> String {
        Bundle.main.object(forInfoDictionaryKey: key) as? String ?? ""
    }
}
